eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/RawTxPublisher.scala

```diff
@@ -0,0 +1,190 @@
+
+/**
+ * Created by t-bast on 10/06/2021.
+ */
+
+/**
+ * This actor publishes a raw transaction without modifying it.
+ * It waits for confirmation or failure before reporting back to the requesting actor.
+ */
+object RawTxPublisher {
+
+  // @formatter:off
+  sealed trait Command
+  case class Publish(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishRawTx) extends Command
+  private case object TimeLocksOk extends Command
+  private case object CheckParentTx extends Command
+  private case object ParentTxOk extends Command
+  private case object ParentTxMissing extends Command
+  private case class WrappedTxResult(result: MempoolTxMonitor.TxResult) extends Command
+  private case class UnknownFailure(reason: Throwable) extends Command
+  case object Stop extends Command
+  // @formatter:on
+
+  def apply(nodeParams: NodeParams, bitcoinClient: ExtendedBitcoinClient, watcher: ActorRef[ZmqWatcher.Command], loggingInfo: TxPublishInfo): Behavior[Command] = {
+    Behaviors.setup { context =>
+      Behaviors.withTimers { timers =>
+        Behaviors.withMdc(loggingInfo.mdc()) {
+          new RawTxPublisher(nodeParams, bitcoinClient, watcher, context, timers, loggingInfo).start()
+        }
+      }
+    }
+  }
+
+}
+
+private class RawTxPublisher(nodeParams: NodeParams,
+                             bitcoinClient: ExtendedBitcoinClient,
+                             watcher: ActorRef[ZmqWatcher.Command],
+                             context: ActorContext[RawTxPublisher.Command],
+                             timers: TimerScheduler[RawTxPublisher.Command],
+                             loggingInfo: TxPublishInfo)(implicit ec: ExecutionContext = ExecutionContext.Implicits.global) {
+
+  import RawTxPublisher._
+
+  private val log = context.log
+
+  def start(): Behavior[Command] = {
+    Behaviors.receiveMessagePartial {
+      case Publish(replyTo, cmd) => checkTimeLocks(replyTo, cmd)
+      case Stop => Behaviors.stopped
+    }
+  }
+
+  def checkTimeLocks(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishRawTx): Behavior[Command] = {
+    val timeLocksChecker = context.spawn(TxTimeLocksMonitor(nodeParams, watcher, loggingInfo), "time-locks-monitor")
+    timeLocksChecker ! CheckTx(context.messageAdapter[TxTimeLocksMonitor.TimeLocksOk](_ => TimeLocksOk), cmd.tx, cmd.desc)
+    Behaviors.receiveMessagePartial {
+      case TimeLocksOk => checkParentPublished(replyTo, cmd)
+      case Stop =>
+        timeLocksChecker ! TxTimeLocksMonitor.Stop
+        Behaviors.stopped
+    }
+  }
+
+  def checkParentPublished(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishRawTx): Behavior[Command] = {
+    cmd.parentTx_opt match {
+      case Some(parentTx) =>
+        context.self ! CheckParentTx
+        Behaviors.receiveMessagePartial {
+          case CheckParentTx =>
+            context.pipeToSelf(bitcoinClient.getTxConfirmations(parentTx.txid)) {
```

Comment:
Why not use `WatchParentTxConfirmed` like in `TxTimeLocksMonitor`? Probably because it wouldn't give you the granularity needed.

Reply:
Yes, and because here we know the parent should appear in the mempool very soon, so it's lighter to check ourselves than go through the watcher (which makes several calls to bitcoin core to fetch more information than we need).